Nil Pointer Deference happening when spec.coreTemplate.spec.replicas is not defined in EMQX CRD

**Describe the bug**

The EMQX CRD version `v2beta1` can cause a nil pointer dereference in the operator controller when updating the `Replicas` status if `spec.coreTemplate.spec.replicas` is not defined in the CRD.

**Logs**

```log
"level":"info","ts":"2025-01-12T17:42:42Z","msg":"Starting workers","controller":"rebalance","controllerGroup":"apps.emqx.io","controllerKind":"Rebalance","worker count":1}
{"level":"info","ts":"2025-01-12T17:42:42Z","msg":"Starting workers","controller":"emqx","controllerGroup":"apps.emqx.io","controllerKind":"EMQX","worker count":1}
{"level":"info","ts":"2025-01-12T17:42:43Z","msg":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference","controller":"emqx","controllerGroup":"apps.emqx.io","controllerKind":"EMQX","EMQX":{"name":"emqx","namespace":"emqx"},"namespace":"emqx","name":"emqx","reconcileID":"eaeaed64-f44c-420c-81e2-8e1613bbeb4c"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x177dc84]
goroutine 248 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x195d980?, 0x2c00a90?})
    /usr/local/go/src/runtime/panic.go:770 +0x132
github.com/emqx/emqx-operator/controllers/apps/v2beta1.(*updateStatus).reconcile(0xc00049e020?, {0x1e76b08?, 0xc00098a690?}, {{0xc0001c5830?, 0xc00058cc08?}, 0xc00098a690?}, 0xc00058cc08?, {0x0?, 0x0?})
    /workspace/controllers/apps/v2beta1/update_emqx_status.go:24 +0x44
github.com/emqx/emqx-operator/controllers/apps/v2beta1.(*EMQXReconciler).Reconcile(0xc000429f50, {0x1e76b08, 0xc00098a690}, {{{0xc0009afaac?, 0x5?}, {0xc0009afaa8?, 0xc0006a9d10?}}})
    /workspace/controllers/apps/v2beta1/emqx_controller.go:137 +0x7c3
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1e7b088?, {0x1e76b08?, 0xc00098a690?}, {{{0xc0009afaac?, 0xb?}, {0xc0009afaa8?, 0x0?}}})
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0002f6960, {0x1e76b40, 0xc0000507d0}, {0x1a0de80, 0xc0000ac4e0})
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:316 +0x3bc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0002f6960, {0x1e76b40, 0xc0000507d0})
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:266 +0x1be
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 145
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:223 +0x50c
```

**To Reproduce**

1. Deploy the EMQX Operator.
2. Create an EMQX CRD without defining `spec.coreTemplate.spec.replicas`:

   ```yaml
   ---
   apiVersion: apps.emqx.io/v2beta1
   kind: EMQX
   metadata:
     name: emqx
     namespace: emqx
   spec:
     image: emqx:5
     metadata:
       labels:
         app: emqx
   ```

3. Observe that the operator crashes due to a nil pointer dereference.

**Expected Behavior**
The EMQX CRD should be deployed successfully without causing a crash in the operator controller. The operator should handle the absence of `replicas` gracefully.

The following manifest resolves the issue:

```yaml
apiVersion: apps.emqx.io/v2beta1
kind: EMQX
metadata:
  name: emqx
  namespace: emqx
spec:
  image: emqx:5
  coreTemplate:
    spec:
    replicas: 1
```

**Additional Information**
The issue is likely caused by the lack of a nil check on `instance.Spec.CoreTemplate.Spec.Replicas` in the controller code, specifically at [[this line](https://github.com/emqx/emqx-operator/blob/2f4089d08b78b91a3cf8aab915a221ece6999e47/controllers/apps/v2beta1/update_emqx_status.go#L24)](<https://github.com/emqx/emqx-operator/blob/2f4089d08b78b91a3cf8aab915a221ece6999e47/controllers/apps/v2beta1/update_emqx_status.go#L24>). A nil pointer dereference occurs when attempting to access the `Replicas` field if it is not specified.

To fix this:

- The controller should ensure that `instance.Spec.CoreTemplate.Spec.Replicas` is checked for `nil` before dereferencing.
- Alternatively, an admission webhook could be added to ensure that a valid value for `replicas` is provided during CRD creation.

**Environment Details:**

- Kubernetes version: 1.32.0
- Cloud provider/provisioner: Self Hosted Kubernetes Cluster
- EMQX Operator version: 2.2.26
- Installation method: Helm via CICD

--- a/api/v2/emqx_types_spec.go
+++ b/api/v2/emqx_types_spec.go
@@ -159,6 +159,7 @@ type EMQXCoreTemplate struct {
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 	// Specification of the desired state of a core node.
 	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +kubebuilder:default={}
 	Spec EMQXCoreTemplateSpec `json:"spec,omitempty"`
 }
 

--- a/config/crd/bases/apps.emqx.io_emqxes.yaml
+++ b/config/crd/bases/apps.emqx.io_emqxes.yaml
@@ -154,6 +154,7 @@ spec:
                         type: string
                     type: object
                   spec:
+                    default: {}
                     description: |-
                       Specification of the desired state of a core node.
                       More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status

--- a/internal/controller/suite_test.go
+++ b/internal/controller/suite_test.go
@@ -34,6 +34,7 @@ import (
 
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
 	ctrl "sigs.k8s.io/controller-runtime"
@@ -156,6 +157,54 @@ var _ = AfterSuite(func() {
 	_ = testEnv.Stop()
 	// err := testEnv.Stop()
 	// Expect(err).NotTo(HaveOccurred())
+})
+
+var _ = Describe("CRD Defaults", Ordered, func() {
+	var ns *corev1.Namespace
+
+	BeforeAll(func() {
+		ns = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "api-defaults-test-ns",
+			},
+		}
+		Expect(k8sClient.Create(ctx, ns)).To(Succeed())
+	})
+
+	AfterAll(func() {
+		Expect(k8sClient.Delete(ctx, ns)).To(Succeed())
+	})
+
+	It("defaults coreTemplate.spec when spec is missing", func() {
+		instance := &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": crdv2.GroupVersion.String(),
+				"kind":       "EMQX",
+				"metadata": map[string]interface{}{
+					"name":      "emqx",
+					"namespace": ns.Name,
+				},
+				"spec": map[string]interface{}{
+					"image": "emqx",
+					"coreTemplate": map[string]interface{}{
+						"metadata": map[string]interface{}{},
+					},
+				},
+			},
+		}
+		Expect(k8sClient.Create(ctx, instance)).To(Succeed())
+
+		actual := &crdv2.EMQX{}
+		Expect(k8sClient.Get(ctx, client.ObjectKey{Namespace: ns.Name, Name: "emqx"}, actual)).To(Succeed())
+		Expect(actual.Spec.CoreTemplate.Spec).To(HaveField("Replicas", And(
+			Not(BeNil()),
+			HaveValue(BeEquivalentTo(2)),
+		)))
+		Expect(actual.Spec.CoreTemplate.Spec).To(HaveField("PodSecurityContext", And(
+			Not(BeNil()),
+			HaveValue(HaveField("RunAsUser", HaveValue(BeEquivalentTo(1000)))),
+		)))
+	})
 })
 
 func newReconcileRound() *reconcileRound {

--- a/internal/controller/update_emqx_status.go
+++ b/internal/controller/update_emqx_status.go
@@ -11,6 +11,7 @@ import (
 	appsv1 "k8s.io/api/apps/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
 )
 
 type updateStatus struct {
@@ -20,9 +21,9 @@ type updateStatus struct {
 func (u *updateStatus) reconcile(r *reconcileRound, instance *crdv2.EMQX) subResult {
 	status := &instance.Status
 
-	status.CoreNodesStatus.Replicas = *instance.Spec.CoreTemplate.Spec.Replicas
+	status.CoreNodesStatus.Replicas = ptr.Deref(instance.Spec.CoreTemplate.Spec.Replicas, 1)
 	if instance.Spec.ReplicantTemplate != nil {
-		status.ReplicantNodesStatus.Replicas = *instance.Spec.ReplicantTemplate.Spec.Replicas
+		status.ReplicantNodesStatus.Replicas = ptr.Deref(instance.Spec.ReplicantTemplate.Spec.Replicas, 1)
 	}
 
 	currentCoreSet, updateCoreSet := switchCoreSet(r, instance)

